src/ufuncs.c

```diff
@@ -340,25 +339,47 @@ static void Yd_Y_func(char **args, npy_intp *dimensions,
                       npy_intp *steps, void *data)
 {
     FuncGEOS_Yd_Y *func = (FuncGEOS_Yd_Y *)data;
-    GEOSGeometry *in1, *ret_ptr;
+    GEOSGeometry *in1;
 
-    GEOS_INIT;
+    // allocate a temporary array to store output GEOSGeometry objects
+    GEOSGeometry **geom_arr = malloc(sizeof(void *) * dimensions[0]);
 
-    BINARY_LOOP {
-        /* get the geometry: return on error */
-        if (!get_geom(*(GeometryObject **)ip1, &in1)) { errstate = PGERR_NOT_A_GEOMETRY; goto finish; }
-        double in2 = *(double *)ip2;
-        if ((in1 == NULL) | (npy_isnan(in2))) {
-            ret_ptr = NULL;
-        } else {
-            ret_ptr = func(ctx, in1, in2);
-            if (ret_ptr == NULL) { errstate = PGERR_GEOS_EXCEPTION; goto finish; }
+    GEOS_INIT_THREADS;
+
+    if (geom_arr != NULL) {
+        BINARY_LOOP {
+            // get the geometry: return on error
+            if (!get_geom(*(GeometryObject **)ip1, &in1)) {
+                errstate = PGERR_NOT_A_GEOMETRY;
+                destroy_geom_arr(ctx, geom_arr, i - 1);
+                break;
+            }
+            double in2 = *(double *)ip2;
+            if ((in1 == NULL) | (npy_isnan(in2))) {
+                // in case of a missing value: return NULL (None)
+                geom_arr[i] = NULL;
+            } else {
+                geom_arr[i] = func(ctx, in1, in2);
+                if (geom_arr[i] == NULL) {
+                    errstate = PGERR_GEOS_EXCEPTION;
+                    destroy_geom_arr(ctx, geom_arr, i - 1);
+                    break;
+                }
+            }
         }
-        OUTPUT_Y;
+    } else {
+        errstate = PGERR_NO_MALLOC;
```

Comment:
@caspervdw what do you think about moving this up below line 345, with a return immediately after?  Something like:
```
GEOSGeometry **geom_arr = malloc(sizeof(void *) * dimensions[0]);
if (geom_arr == NULL) {
    errstate = PGERR_NO_MALLOC;
    return
}
```

Then the "happy path" becomes simpler:
```
GEOS_INIT_THREADS;
BINARY_LOOP { ... }
GEOS_FINISH_THREADS;

if (errstate == PGERR_SUCCESS) {
    geom_arr_to_npy(geom_arr, args[2], steps[2], dimensions[0]);
}
free(geom_arr);
```

Reply:
Thanks for the suggestion! The `errstate` variable doesn't exist outside of the `GEOS_INIT` and `GEOS_FINISH`, but I'll just set the python errorstate directly and return instead.